google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java

```diff
@@ -0,0 +1,251 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub.metrics;
+
+import com.google.api.gax.tracing.ApiTracerFactory;
+import com.google.api.gax.tracing.SpanName;
+import com.google.cloud.bigtable.stats.StatsRecorderWrapper;
+import com.google.cloud.bigtable.stats.StatsWrapper;
+import com.google.common.base.Stopwatch;
+import java.util.Map;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * A {@link BigtableTracer} that records built-in metrics and publish under the
+ * bigtable.googleapis.com/client namespace
+ */
+class BuiltinMetricsTracer extends BigtableTracer {
+
+  private final StatsRecorderWrapper recorder;
+
+  private final ApiTracerFactory.OperationType operationType;
+  private final SpanName spanName;
+
+  // Operation level metrics
+  private final AtomicBoolean opFinished = new AtomicBoolean();
+  private final Stopwatch operationTimer = Stopwatch.createStarted();
+  private final Stopwatch firstResponsePerOpTimer = Stopwatch.createStarted();
+
+  // Attempt level metrics
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private volatile int attempt = 0;
+
+  // Total application latency
+  // Total application latency needs to be atomic because it's accessed from different threads. E.g.
+  // request() from
+  // user thread and attempt failed from grpc thread.
+  private final AtomicLong totalApplicationLatency = new AtomicLong(0);
+  // Stopwatch is not thread safe so this is a workaround to check if the stopwatch changes is
+  // flushed to memory.
+  private final AtomicBoolean applicationLatencyTimerIsRunning = new AtomicBoolean();
+  private final Stopwatch applicationLatencyTimer = Stopwatch.createUnstarted();
+
+  // Monitored resource labels
+  private String tableId = "undefined";
+  private String zone = "undefined";
+  private String cluster = "undefined";
+
+  // gfe stats
+  private AtomicLong gfeMissingHeaders = new AtomicLong(0);
+
+  BuiltinMetricsTracer(
+      ApiTracerFactory.OperationType operationType,
+      SpanName spanName,
+      Map<String, String> attributes,
+      StatsWrapper statsWrapper,
+      @Nullable StatsRecorderWrapper statsRecorderWrapper) {
+    this.operationType = operationType;
+    this.spanName = spanName;
+    if (statsRecorderWrapper != null) {
+      // A workaround for test to pass in a mock StatsRecorderWrapper
+      this.recorder = statsRecorderWrapper;
+    } else {
+      this.recorder = new StatsRecorderWrapper(operationType, spanName, attributes, statsWrapper);
+    }
```

Comment:
I think you should add a factory method and avoid if statements for test behavior. Also, if add the factory method to statsWrapper like i mentioned in the other PR, it will look something liek this:

```java
static BuiltinMetricsTracer create(OperationType, SpanName, ...) {
   return new BuiltinMetricsTracer(opType, spanName, attribs, statsWrapper.createRecorder(...));
}
// VisibleForTesting
BuiltinMetricsTracer(OperationType, SpanName, ..., StatsRecorderWrapper recorder) {
}
```